svcdiscovery/plugin/svcdiscovery/plugin.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2020 SUSE
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package svcdiscovery
+
+import (
+	"context"
+	"net"
+
+	"github.com/coredns/coredns/plugin"
+	clog "github.com/coredns/coredns/plugin/pkg/log"
+	"github.com/coredns/coredns/request"
+	"github.com/miekg/dns"
+)
+
+// ServiceDiscovery is a CoreDNS plugin for handling Cloud Foundry App Service
+// Discovery.
+type ServiceDiscovery struct {
+	Next plugin.Handler
+
+	log       clog.P
+	sdcClient *SDCClient
+	ttl       uint32
+}
+
+// Name satisfies plugin.Handler.Name.
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names. It satisfies satisfies
+// plugin.Handler.ServeDNS.
+func (sd *ServiceDiscovery) ServeDNS(
+	ctx context.Context,
+	rw dns.ResponseWriter,
+	req *dns.Msg,
+) (int, error) {
+	qclass := req.Question[0].Qclass
+	qtype := req.Question[0].Qtype
+	qname := req.Question[0].Name
+
+	if qclass == dns.ClassINET && (qtype == dns.TypeA || qtype == dns.TypeAAAA) {
+		ips, err := sd.sdcClient.Discover(qname)
+		if err != nil {
+			sd.log.Error(err)
+			return dns.RcodeServerFailure, err
+		}
+
+		if len(ips) > 0 {
+			if err := sd.respond(rw, req, ips); err != nil {
+				return dns.RcodeServerFailure, err
+			}
+			return dns.RcodeSuccess, nil
+		}
+	}
+
+	return plugin.NextOrFailure(pluginName, sd.Next, ctx, rw, req)
+}
+
+func (sd *ServiceDiscovery) respond(rw dns.ResponseWriter, req *dns.Msg, ips []string) error {
+	state := request.Request{W: rw, Req: req}
+	qtype := state.QType()
+	qname := state.Name()
+
+	res := &dns.Msg{}
+	res.SetReply(req)
+	res.Authoritative = true
+
+	answer := make([]dns.RR, 0)
```

Comment:
```suggestion
	answer := make([]dns.RR, 0, len(ips))
```
No reason not to allocate ahead of time; since the capacity is separate from the length, it's not an issue if we end up not using some IPs.

Reply:
Hmmmm... while the logic doesn't guarantee it, I guess it will always be the case. I thought about it and decided to leave it out, but I'll take it.